src/precice/impl/ParticipantImpl.cpp

```diff
@@ -685,7 +724,26 @@ int ParticipantImpl::getMeshVertexSize(
                 meshName, _accessor->getName());
   MeshContext &context = _accessor->usedMeshContext(meshName);
   PRECICE_ASSERT(context.mesh.get() != nullptr);
-  return context.mesh->nVertices();
+
+  // Returns true if we have api access configured and we run in parallel and have a received mesh
+  if ((context.userDefinedAccessRegion || requiresUserDefinedAccessRegion(meshName)) && _accessor->isDirectAccessAllowed(meshName)) {
+    // filter nVertices to the actual number of vertices queried by the user
+    PRECICE_CHECK(context.userDefinedAccessRegion, "The function getMeshVertexSize was called on the received mesh \"{0}\", "
+                                                   "but no access region was defined although this is necessary for parallel runs. "
+                                                   "Please define an access region using \"setMeshAccessRegion()\" before calling \"getMeshVertexSize()\".",
+                  meshName);
+    auto result = std::count_if(context.mesh->vertices().cbegin(), context.mesh->vertices().cend(),
+                                [bb = context.userDefinedAccessRegion](const auto &v) { return bb->contains(v); });
```

Comment:
Let's start moving these to `mesh/Utils.hpp`.

```suggestion
    auto result = mesh::countInBB(context.mesh, context.userDefinedAccessRegion);
```